source/linear_algebra/matrix.cpp

```diff
@@ -4,83 +4,83 @@
     }
-    if ((zy < (that.zy - test_precision)) || (zy > (that.zy + test_precision))) {
-    return false;
+    
+    Vector Matrix::operator*(const Vector & that) const {
+        return Vector((xx * that.x) + (xy * that.y) + (xz * that.z), (yx * that.x) + (yy * that.y) + (yz * that.z), (zx * that.x) + (zy * that.y) + (zz * that.z));
     }
-    if ((zz < (that.zz - test_precision)) || (zz > (that.zz + test_precision))) {
-    return false;
+    
+    Vector operator*(Vector that, const Matrix & those) {
+        return Vector((that.x * those.xx) + (that.y * those.yx) + (that.z * those.zx), (that.x * those.xy) + (that.y * those.yy) + (that.z * those.zy), (that.x * those.xz) + (that.y * those.yz) + (that.z * those.zz));
     }
-    return true;
-}
-
-Matrix Matrix::operator+(const Matrix & that) const {
-    return Matrix(this->xx + that.xx, this->xy + that.xy, this->xz + that.xz, this->yx + that.yx, this->yy + that.yy, this->yz + that.yz, this->zx + that.zx, this->zy + that.zy, this->zz + that.zz);
-}
-
-Matrix Matrix::operator-(const Matrix & that) const {
-    return Matrix(this->xx - that.xx, this->xy - that.xy, this->xz - that.xz, this->yx - that.yx, this->yy - that.yy, this->yz - that.yz, this->zx - that.zx, this->zy - that.zy, this->zz - that.zz);
-}
-
-Matrix Matrix::operator*(const float & that) const {
-    return Matrix(xx * that, xy * that, xz * that, yx * that, yy * that, yz * that, zx * that, zy * that, zz * that);
-}
-
-Matrix Matrix::operator*(const Matrix & that) const {
-    return Matrix((xx * that.xx) + (xy * that.yx) + (xz * that.zx), (xx * that.xy) + (xy * that.yy) + (xz * that.zy), (xx * that.xz) + (xy * that.yz) + (xz * that.zz),
-                  (yx * that.xx) + (yy * that.yx) + (yz * that.zx), (yx * that.xy) + (yy * that.yy) + (yz * that.zy), (yx * that.xz) + (yy * that.yz) + (yz * that.zz),
-                  (zx * that.xx) + (zy * that.yx) + (zz * that.zx), (zx * that.xy) + (zy * that.yy) + (zz * that.zy), (zx * that.xz) + (zy * that.yz) + (zz * that.zz));
-}
-
-cpp_utils::linalg::Matrix cpp_utils::linalg::operator*(float that, const cpp_utils::linalg::Matrix & those) {
-    return those * that;
-}
-
-Vector Matrix::operator*(const Vector & that) const {
-    return Vector((xx * that.x) + (xy * that.y) + (xz * that.z), (yx * that.x) + (yy * that.y) + (yz * that.z), (zx * that.x) + (zy * that.y) + (zz * that.z));
-}
-
-cpp_utils::linalg::Vector cpp_utils::linalg::operator*(cpp_utils::linalg::Vector that, const cpp_utils::linalg::Matrix & those) {
-    return Vector((that.x * those.xx) + (that.y * those.yx) + (that.z * those.zx), (that.x * those.xy) + (that.y * those.yy) + (that.z * those.zy), (that.x * those.xz) + (that.y * those.yz) + (that.z * those.zz));
-}
-
-Matrix Matrix::inverse() const {
-    if(this->det() == 0) {
-        return Matrix(0, 0, 0, 0, 0, 0, 0, 0, 0);
+    
+    Matrix Matrix::inverse() const {
+        if(this->det() == 0) {
+            return Matrix(0, 0, 0, 0, 0, 0, 0, 0, 0);
```

Comment:
I would prefer an error message here but I don't know how to do them.

Reply:
Throw new std::exception("exception message")

Reply:
I think it works now including unit tests.